src/jail/linux/dns.rs

```diff
@@ -0,0 +1,204 @@
+                        debug!("Sent dummy DNS response to {}", src);
+                    }
+                }
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                continue;
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::TimedOut => {
+                continue;
+            }
+            Err(e) => {
+                if !shutdown.load(Ordering::Relaxed) {
+                    error!("DNS server receive error: {}", e);
+                }
+                break;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn build_dummy_response(query: &[u8]) -> Result<Vec<u8>> {
+    if query.len() < 12 {
+        anyhow::bail!("DNS query too short");
+    }
+
+    let mut response = Vec::with_capacity(512);
+
+    response.extend_from_slice(&query[0..2]);
+
+    let mut flags = ((query[2] as u16) << 8) | (query[3] as u16);
+    flags |= 0x8000;
+    flags &= !0x7800;
+    flags &= !0x000F;
+
+    response.push((flags >> 8) as u8);
+    response.push((flags & 0xFF) as u8);
+
+    // Copy question count from query
+    response.extend_from_slice(&query[4..6]);
+
+    response.push(0);
+    response.push(1);
+
+    response.push(0);
+    response.push(0);
+    response.push(0);
+    response.push(0);
+
+    let query_end = find_query_end(query, 12)?;
+    response.extend_from_slice(&query[12..query_end]);
+
+    response.push(0xC0);
+    response.push(0x0C);
+
+    response.push(0);
+    response.push(1);
+
+    response.push(0);
+    response.push(1);
+
+    response.push(0);
+    response.push(0);
+    response.push(0);
+    response.push(60);
+
+    response.push(0);
+    response.push(4);
+
+    response.extend_from_slice(&DUMMY_IPV4.octets());
+
+    Ok(response)
+}
+
+fn find_query_end(packet: &[u8], start: usize) -> Result<usize> {
+    let mut pos = start;
+
+    while pos < packet.len() {
+        let len = packet[pos] as usize;
+        if len == 0 {
+            pos += 1;
+            break;
+        }
+        if len >= 0xC0 {
+            pos += 2;
+            break;
+        }
+        pos += len + 1;
+    }
+
+    if pos + 4 > packet.len() {
+        anyhow::bail!("Malformed DNS query");
+    }
+
+    Ok(pos + 4)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_dns_response_builder() {
+        let query = vec![
+            0x12, 0x34, 0x01, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x06, b'g',
+            b'o', b'o', b'g', b'l', b'e', 0x03, b'c', b'o', b'm', 0x00, 0x00, 0x01, 0x00, 0x01,
+        ];
+
+        let response = build_dummy_response(&query).unwrap();
+
+        assert_eq!(response[0..2], query[0..2]);
+
+        assert_eq!(response[2] & 0x80, 0x80);
+
+        let answer_count = ((response[6] as u16) << 8) | (response[7] as u16);
+        assert_eq!(answer_count, 1);
+
+        let response_ip = &response[response.len() - 4..];
+        assert_eq!(response_ip, &[6, 6, 6, 6]);
+    }
+}
```

Comment:
Investigate using an off-the-shelf library for DNS servers in Rust or at least a package that provides protocol parsing for us.